client/tests/integration/smartcontracts/validator_with_admin/src/lib.rs

```diff
@@ -4,18 +4,47 @@
 #![no_std]
 #![no_main]
 
+use iroha_validator::{parse, prelude::*};
+
 #[cfg(not(test))]
 extern crate panic_halt;
 
-use iroha_validator::{parse, prelude::*};
+/// Default validator
+///
+/// You can put state in your validator and use that state
+struct Validator;
 
-/// Allow operation if authority is `admin@admin` and fallback to
-/// [`iroha_default_validator::validate()`] if not.
+impl DefaultValidator for Validator {}
```

Comment:
Also if you don't use `iroha_default_validator` it should be removed from `Cargo.toml`

Reply:
> From where did `DefaultValidator` come? I thought it should be imported from `iroha_default_validator`

That was not possible. Due to orphan rule I cannot have `impl<T: DefaultValidator> Validate for T {}` implemented in `iroha_default_validator`

> Also if you don't use iroha_default_validator it should be removed from Cargo.toml

makes sense

Reply:
Then what about

```rust
//! Inside `iroha_default_validator`

pub trait DefaultValidate {
    // Same functions as in `Validate` but with our default implementation
}

pub struct DefaultValidator<T: DefaultValidate>(T);

impl<T: DefaultValidate> Validate for DefaultValidator<T> {
    // Fallback to `DefaultValidate` impls
}
```

So that user could write:

```rust
pub struct MyValidator;

impl DefaultValidate for MyValidator {
    // Override the functions user wants
}

pub fn entypoint(/* ...  */) -> Verdict {
    let validator = iroha_default_validator::DefaultValidator(MyValidator);
    // ...
}
```

Btw user code can look a bit confusing because of word `Default` here, but you get the idea

Reply:
IMO you shouldn't be overriding parts and pieces of the default validator. It's what it is, default is the default, etc. Also it would require a user to include dependency on `iroha_default_validator` crate, which is kinda silly because it's supposed to be an executable, not library.

But look at this. You can still define a new validator that sometimes delegates to the default validator. You do this via newtype wrapper around the default validator. Here as a full example of how you can achieve overriding of default validation in your own custom validator:
```rs
//! Custom validator.

#![no_std]
#![no_main]

use iroha_validator::prelude::*;

#[cfg(not(test))]
extern crate panic_halt;

/// Default validator
///
/// Use this validator to fallback to default validation
struct DefaultValidator;
impl iroha_validator::DefaultValidator for DefaultValidator {}

/// Validator
///
/// You can put state in your validator and use that state
struct Validator(DefaultValidator);

impl Validate for Validator {
    fn validate_transaction(
        &mut self,
        authority: &AccountId,
        transaction: SignedTransaction,
    ) -> Verdict {
        self.0.validate_transaction(authority, transaction)
    }
    fn validate_instruction(
        &mut self,
        authority: &AccountId,
        instruction: &InstructionBox,
    ) -> Verdict {
        self.0.validate_instruction(authority, instruction)
    }
    fn validate_query(&mut self, authority: &AccountId, query: &QueryBox) -> Verdict {
        self.0.validate_query(authority, query)
    }
}

/// Validate operation.
#[entrypoint(params = "[authority, operation]")]
pub fn validate(
    authority: <Account as Identifiable>::Id,
    operation: NeedsValidationBox,
) -> Verdict {
    let mut validator = Validator(DefaultValidator);

    match operation {
        // NOTE: Invoked from Iroha
        NeedsValidationBox::Transaction(transaction) => {
            validator.validate_transaction(&authority, transaction)
        }

        // NOTE: Invoked only from another Wasm
        NeedsValidationBox::Instruction(instruction) => {
            let verdict = validator.validate_instruction(&authority, &instruction);

            if !verdict.is_deny() {
                instruction.execute();
            }

            verdict
        }

        // NOTE: Invoked only from another Wasm
        NeedsValidationBox::Query(query) => validator.validate_query(&authority, &query),
    }
}
```

Reply:
I've found a way to move the default code back into `default_validator`. Check the test `validator_with_admin` to see how to override the `DefaultValidator` (it has been made into a struct now)